src/runtime-rs/crates/hypervisor/src/device/pci_path.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2019-2023 Alibaba Cloud
+// Copyright (c) 2019-2023 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use anyhow::{anyhow, Context, Result};
+use lazy_static::lazy_static;
+use regex::Regex;
+
+// get pci path from bdf, and the bdf will be start with "0000:00"
+lazy_static! {
+    static ref RE: Regex = Regex::new(r"^0000:00:([[:xdigit:]]{2}).[[:xdigit:]]$").unwrap();
+}
+
+// Tips:
+// The Re-write `PciSlot` and `PciPath` with rust that it origins from `pcipath.go`:
+//
+
+// The PCI spec reserves 5 bits for slot number (a.k.a. device
+// number), giving slots 0..31
+const PCI_SLOT_BITS: u32 = 5;
+const MAX_PCI_SLOTS: u32 = (1 << PCI_SLOT_BITS) - 1;
+
+// A PciSlot describes where a PCI device sits on a single bus
+//
+// This encapsulates the PCI slot number a.k.a device number, which is
+// limited to a 5 bit value [0x00..0x1f] by the PCI specification
+//
+// To support multifunction device's, It's needed to extend
+// this to include the PCI 3-bit function number as well.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct PciSlot(pub u8);
+
+impl PciSlot {
+    pub fn convert_from_string(s: &str) -> Result<PciSlot> {
```

Comment:
It would be more idiomatic to implement `TryFrom` for `PciSlot`:

```rust
use std::convert::TryFrom;

impl TryFrom<&str> for PciSlot {
    type Error = anyhow::Error;

    fn try_from(s: &str) -> Result<Self> {
      // ...
    }
}

impl TryFrom<u32> for PciSlot {
    type Error = anyhow::Error;

    fn try_from(slot: u32) -> Result<Self> {
      // ...
    }
}
```

Reply:
fixed